File: dataset/conv_metis_to_mtx.py
import sys

def convert_metis_to_mtx(metis_file, mtx_file):
    print(f"Reading {metis_file}...")
    
    with open(metis_file, 'r') as f:
        # Read header
        line = f.readline()
        parts = line.split()
        num_vertices = int(parts[0])
        num_edges = int(parts[1])
        
        print(f"Graph dimensions: {num_vertices} vertices, {num_edges} edges")
        
        # Prepare adjacency list
        adj = [[] for _ in range(num_vertices + 1)]
        
        for i in range(1, num_vertices + 1):
            line = f.readline().strip()
            if line:
                neighbors = list(map(int, line.split()))
                adj[i].extend(neighbors)
    
    print(f"Writing {mtx_file}...")
    
    with open(mtx_file, 'w') as f:
        # Write MTX header
        f.write("%%MatrixMarket matrix coordinate pattern symmetric\n")
        f.write(f"{num_vertices} {num_vertices} {num_edges}\n")
        
        # Write edges (u,v) for each neighbor v of u
        for u in range(1, num_vertices + 1):
            for v in adj[u]:
                if u < v:  # To avoid duplicates in symmetric format
                    f.write(f"{v} {u}\n")

    print("Conversion complete.")
    
    
if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python convert_metis_to_mtx.py <input.metis> <output.mtx>")
        sys.exit(1)
    
    metis_file = sys.argv[1]
    mtx_file = sys.argv[2]
    
    convert_metis_to_mtx(metis_file, mtx_file)

File: dataset/conv_metis_to_mtx.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <fstream>
#include <sstream>
#include <iostream>
#include <set>
using namespace std;

/*
METIS graph format to Matrix Market (.mtx) converter.

METIS format:
  % comment lines (ignored)
  <num_vertices> <num_edges> [fmt] [ncon]   (header)
  <adj1> <adj2> ...                          (una riga per vertice, 1-based)
  ...

Matrix Market format:
  %%MatrixMarket matrix coordinate pattern symmetric
  <num_rows> <num_cols> <num_entries>   (num_entries = num_edges, upper triangle)
  <row> <col>                           (1-based, row <= col)
*/

int main(int argc, char *argv[])
{
    if (argc < 3)
    {
        cerr << "Usage: " << argv[0] << " <input_metis> <output_mtx>" << endl;
        return 1;
    }

    string input = argv[1], output = argv[2];
    ifstream infile(input);

    if (!infile.is_open())
    {
        cerr << "Error opening input file: " << input << endl;
        return 1;
    }

    // Leggi header saltando commenti
    string line;
    int num_vertices = 0, num_edges = 0;
    while (getline(infile, line))
    {
        if (line.empty()) continue;
        if (line[0] == '%') continue; // Commenti METIS

        istringstream iss(line);
        if (!(iss >> num_vertices >> num_edges))
        {
            cerr << "Error parsing METIS header." << endl;
            return 1;
        }
        break; // Header letto, esci
    }

    if (num_vertices == 0)
    {
        cerr << "Empty or invalid METIS file." << endl;
        return 1;
    }

    // Leggi lista adiacenza e deduplicat in upper triangle
    set<pair<int,int>> edges;
    int vertex = 1;
    while (getline(infile, line))
    {
        if (line.empty())
        {
            // Riga vuota = nodo senza archi, avanza comunque
            vertex++;
            continue;
        }
        if (line[0] == '%') continue;

        istringstream iss(line);
        int adj;
        while (iss >> adj)
        {
            if (adj == vertex) continue; // salta self-loop
            // Salva solo upper triangle (row <= col)
            int r = min(vertex, adj);
            int c = max(vertex, adj);
            edges.insert({r, c});
        }
        vertex++;

        if (vertex > num_vertices) break;
    }
    infile.close();

    ofstream outfile(output);
    if (!outfile.is_open())
    {
        cerr << "Error opening output file: " << output << endl;
        return 1;
    }

    outfile << "%%MatrixMarket matrix coordinate pattern symmetric\n";
    outfile << "% Converted from METIS format: " << input << "\n";
    outfile << num_vertices << " " << num_vertices << " " << (int)edges.size() << "\n";

    for (const auto& e : edges)
        outfile << e.first << " " << e.second << "\n";

    outfile.close();
    cout << "Done: " << num_vertices << " vertices, " << edges.size() << " edges." << endl;
    return 0;
}